### ISSUE TITLE
"Task exception was never retrieved" reported for a canceled task

### ISSUE THREAD

**Issue by 35afb67f-c593-48f4-8239-6dce86747d79** [python/cpython#74693]
BPO | [30508](https://bugs.python.org/issue30508)
--- | :---
Nosy | @ned-deily, @1st1
PRs | <li>python/cpython#2050</li><li>python/cpython#2109</li><li>python/cpython#2110</li>
Files | <li>[never-retrieved.py](https://bugs.python.org/file46911/never-retrieved.py "Uploaded as text/plain at 2017-05-30.02:14:05 by Miguel Grinberg")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-06-11.14:12:51.714>
created_at = <Date 2017-05-30.02:14:06.380>
labels = ['type-bug', '3.7', 'expert-asyncio']
title = '"Task exception was never retrieved" reported for a canceled task'
updated_at = <Date 2017-06-11.15:42:01.895>
user = 'https://bugs.python.org/MiguelGrinberg'
```

bugs.python.org fields:
```python
activity = <Date 2017-06-11.15:42:01.895>
actor = 'ned.deily'
assignee = 'none'
closed = True
closed_date = <Date 2017-06-11.14:12:51.714>
closer = 'yselivanov'
components = ['asyncio']
creation = <Date 2017-05-30.02:14:06.380>
creator = 'Miguel Grinberg'
dependencies = []
files = ['46911']
hgrepos = []
issue_num = 30508
keywords = []
message_count = 7.0
messages = ['294732', '294733', '295708', '295709', '295711', '295712', '295721']
nosy_count = 3.0
nosy_names = ['ned.deily', 'yselivanov', 'Miguel Grinberg']
pr_nums = ['2050', '2109', '2110']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue30508'
versions = ['Python 3.5', 'Python 3.6', 'Python 3.7']
```

</p></details>

**Comment by 35afb67f-c593-48f4-8239-6dce86747d79** [comment]
I am seeing a strange issue that occurs when a task that is awaiting an asyncio.wait_for() is cancelled. I created a simple example that I think demonstrates the issue, even though it isn't exactly how it manifests on my application.

When I run the attached script never-retrieved.py I get the following error:

```pytb
Task exception was never retrieved
future: <Task finished coro=<crash() done, defined at never-retrieved.py:4> exception=ZeroDivisionError('division by zero',)>
Traceback (most recent call last):
  File "/usr/lib/python3.5/asyncio/tasks.py", line 239, in _step
    result = coro.send(None)
  File "never-retrieved.py", line 5, in crash
    a = 1/0
ZeroDivisionError: division by zero
```


You can see in the script that the future was cancelled, but the cancel() call was made after the task finished in a zero division error. I think the cancel() call should update the internal state of the future so that the "exception was never retrieved" error does not appear.

My application has a more complex setup that I have been unable to reproduce with a simple example. I have a task that is waiting on asyncio.wait_for(fut, timeout), with fut subsequently waiting on a websocket server's receive function. When the websocket client closes the connection, a bunch of cancellations happen, but this future inside the wait_for call crashes before wait_for gets to call cancel() on it. Even though I need to investigate this crash, the fact is that wait_for did cancel this future, but because it already ended in an error the "never retried" error is reported anyway.

**Comment by 1st1** [comment]
Yes, this is a known problem to me, thank you for creating the issue. Will work on a fix soon.

**Comment by 1st1** [comment]
New changeset 7ce1c6fb579a01bb184224a10019039fde9c8eaf by Yury Selivanov in branch 'master':
bpo-30508: Don't log exceptions if Task/Future "cancel()" method called (bpo-2050)
https://github.com/python/cpython/commit/7ce1c6fb579a01bb184224a10019039fde9c8eaf

**Comment by 1st1** [comment]
New changeset 176f2ebdad93f20876c08efabd364a0e6c86de14 by Yury Selivanov in branch '3.6':
bpo-30508: Don't log exceptions if Task/Future "cancel()" method was called. (bpo-2109)
https://github.com/python/cpython/commit/176f2ebdad93f20876c08efabd364a0e6c86de14

**Comment by 1st1** [comment]
New changeset d24c8287e226ac9983caf6bb826a7b53142ee31f by Yury Selivanov in branch '3.5':
bpo-30508: Don't log exceptions if Task/Future "cancel()" method was called. (bpo-2110)
https://github.com/python/cpython/commit/d24c8287e226ac9983caf6bb826a7b53142ee31f

**Comment by 1st1** [comment]
Fixed.  Ned, this will be included in 3.6.2, right?

**Comment by ned-deily** [comment]
Yury, yes (3.6.2 cutoff is about 24 hours from now)
